src/pwm_input.rs

```diff
@@ -0,0 +1,385 @@
+    fn pwm_input<PINS, T>(
+        self,
+        PINS,
+        apb: &mut APB1,
+        mapr: &mut MAPR,
+        dbg: &mut DBG,
+        clocks: &Clocks,
+        mode: Configuration<T>,
+    ) -> Result<PwmInput<Self, PINS>, TimerInitError>
+    where
+        PINS: Pins<Self>,
+        T: Into<Hertz>;
+}
+
+impl PwmInputExt for TIM2 {
+    fn pwm_input<PINS, T>(
+        self,
+        pins: PINS,
+        apb: &mut APB1,
+        mapr: &mut MAPR,
+        dbg: &mut DBG,
+        clocks: &Clocks,
+        mode: Configuration<T>,
+    ) -> Result<PwmInput<Self, PINS>, TimerInitError>
+    where
+        PINS: Pins<Self>,
+        T: Into<Hertz>,
+    {
+        mapr.mapr()
+            .modify(|_, w| unsafe { w.tim2_remap().bits(PINS::REMAP) });
+        tim2(self, pins, clocks, apb, mode, dbg)
+    }
+}
+
+impl PwmInputExt for TIM3 {
+    fn pwm_input<PINS, T>(
+        self,
+        pins: PINS,
+        apb: &mut APB1,
+        mapr: &mut MAPR,
+        dbg: &mut DBG,
+        clocks: &Clocks,
+        mode: Configuration<T>,
+    ) -> Result<PwmInput<Self, PINS>, TimerInitError>
+    where
+        PINS: Pins<Self>,
+        T: Into<Hertz>,
+    {
+        mapr.mapr()
+            .modify(|_, w| unsafe { w.tim3_remap().bits(PINS::REMAP) });
+        tim3(self, pins, clocks, apb, mode, dbg)
+    }
+}
+
+impl PwmInputExt for TIM4 {
+    fn pwm_input<PINS, T>(
+        self,
+        pins: PINS,
+        apb: &mut APB1,
+        mapr: &mut MAPR,
+        dbg: &mut DBG,
+        clocks: &Clocks,
+        mode: Configuration<T>,
+    ) -> Result<PwmInput<Self, PINS>, TimerInitError>
+    where
+        PINS: Pins<Self>,
+        T: Into<Hertz>,
+    {
+        mapr.mapr()
+            .modify(|_, w| w.tim4_remap().bit(PINS::REMAP == 1));
+        tim4(self, pins, clocks, apb, mode, dbg)
+    }
+}
+
+fn get_score(x: u32, y: u32) -> u32 {
+    if x > y {
+        x - y
+    } else {
+        y - x
+    }
+}
+/// F_timer = PCLK / [ (ARR+1) * (PRESC+1) ]
+/// We want maximum resolution, so ARR = u16::MAX;
+/// <=> PRESC = PCLK / [ u16::MAX * F_timer ] -1
+fn compute_arr_presc(freq: u32, clock: u32) -> Result<(u16, u16), TimerInitError> {
+    let mut arr: u32 = (core::u16::MAX - 1) as u32;
+    let mut presc: u32;
+
+    let mut curr_best_presc: u32 = 0;
+    let mut curr_best_arr: u32 = 0;
+    let mut curr_best_distance: u32 = core::u32::MAX;
+
+    for _ in 0..65530 {
+        presc = clock / ((arr + 1) * freq);
+        let computed_freq = clock / ((arr + 1) * (presc + 1));
+        if presc > core::u16::MAX as u32 {
+            break;
+        } else {
+            let score = get_score(computed_freq, freq);
+            if score < curr_best_distance {
+                curr_best_distance = score;
+                curr_best_arr = arr;
+                curr_best_presc = presc;
+            }
+        }
+        arr -= 1;
+    }
+    if curr_best_arr <= core::u16::MAX as u32 || curr_best_presc < core::u16::MAX as u32 {
```

Comment:
I think this condition is bad as you'll convert something over u16::MAX.

Reply:
Thank you, this is now solved !